src/tamp_improv/approaches/rl_improvisational_policy.py

```diff
@@ -0,0 +1,134 @@
+"""RL-based implementation of the improvisational policy."""
+
+from typing import Optional
+
+import numpy as np
+from numpy.typing import NDArray
+from stable_baselines3 import PPO
+from stable_baselines3.common.callbacks import BaseCallback
+
+from tamp_improv.approaches.base_improvisational_tamp_approach import (
+    ImprovisationalPolicy,
+)
+from tamp_improv.benchmarks.pushing_env import PushingEnvWrapper
+
+
+class TrainingProgressCallback(BaseCallback):
+    """Tracks and reports training progress.
+
+    This gets called after each training step to monitor:
+    - Success rate (how often we successfully clear the target area)
+    - Episode lengths (how many steps it takes to succeed)
+    """
+
+    def __init__(self, check_freq: int = 1000, verbose: int = 1):
+        super().__init__(verbose)
+        self.check_freq = check_freq
+        self.success_history: list[bool] = []
+        self.episode_lengths: list[int] = []
+        self.current_length = 0  # Counter for current episode
+
+    def _on_step(self) -> bool:
+        """Called after each training step."""
+        self.current_length += 1
+        # dones indicates if episode ended this step
+        dones = self.locals["dones"]
+        infos = self.locals["infos"]
+
+        if dones[0]:  # If episode ended
+            # Check if it ended due to success (clearing area) vs timeout
+            success = not infos[0].get("TimeLimit.truncated", False)
+            self.success_history.append(success)
+            self.episode_lengths.append(self.current_length)
+            self.current_length = 0
+
+            # Print stats periodically
+            if len(self.success_history) % self.check_freq == 0:
+                recent_successes = self.success_history[-self.check_freq :]
+                recent_lengths = self.episode_lengths[-self.check_freq :]
+                success_rate = sum(recent_successes) / len(recent_successes)
+                avg_length = sum(recent_lengths) / len(recent_lengths)
+                print(f"Episodes: {len(self.success_history)}")
+                print(f"Success rate: {success_rate:.2%}")
+                print(f"Average episode length: {avg_length:.1f}")
+
+        return True  # Continue training
+
+
+class RLImprovisationalPolicy(
+    ImprovisationalPolicy[NDArray[np.float32], NDArray[np.float32]]
+):
+    """The actual RL policy that learns to push block 2 away."""
```

Comment:
why 2 away?

Reply:
This is our manual setting: block 2 is the impediment and we want to move it away from the target area if it's blocking the area, and so afterwards we can move block 1 (our target) into the target area.

Reply:
got it -- but the actual implementation of this class is general and doesn't assume anything about blocks. perhaps change the wording to something like "For example, in the blocks environment, ..."